File: Trees/sum_of_left_leaves.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    
    def sumOfLeftLeaves(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        #base case, current node is None
        if not root:
            return 0
        
        #recursive case 1: left child is a leaf
        #in this case, just return the value of the left child
        #plus the sumOfLeftLeaves() of right child
        if root.left and not root.left.left and not root.left.right:
            return root.left.val + self.sumOfLeftLeaves(root.right)
        
        #recursive case 2: left child is NOT a leaf
        #sum the left leaves of the left subtree and the right subtree
        return self.sumOfLeftLeaves(root.left) + self.sumOfLeftLeaves(root.right)

File: Trees/SumOfLeftLeaves.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int sumOfLeftLeaves(TreeNode root) {
        
        // base case: the node is null
        if (root == null) {
            return 0;
        }
        
        // recursive case #1: left child is a leaf
        // return the value of the left leaf + sumOfLeftLeaves() of the right subtree
        if (root.left != null && root.left.left == null && root.left.right == null) {
            return root.left.val + sumOfLeftLeaves(root.right);
        }
        
        // recursive case #2: left child is not a leaf
        // return the sumOfLeftLeaves() of the left subtree + sumOfLeftLeaves() of the right subtree
        return sumOfLeftLeaves(root.left) + sumOfLeftLeaves(root.right);
        
    }
    
}
